Implement API priorities (reading)

Azure7111 opened the pull request:
https://community.openproject.org/work_packages/17411

Commit:
Add position, isActive, isDefault and self link to priority

Commit:
Removed unnecessary string interpolations

Commit:
Add specs for reading priorities

Commit:
Require view_work_packages permission for priorities access

Commit:
Added single priority endpoint

Commit:
Add priority integration tests

* Error responses
* priorities/:id endpoint

Commit:
Minor codestyle

Commit:
Added priority as link to work_package and changed specs accordingly

Commit:
Added missing properties to examples

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://feature-api-read-priorities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://feature-api-read-priorities.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

myabc commented:
@Azure7111 `rubocop -a` may help quieten Hound :smile_cat:

Commit:
Fix style offenses

NobodysNightmare commented:
@myabc:

```
$ rubocop -a
...
1436 files inspected, 20751 offenses detected, 16793 offenses corrected
$ git status | grep modified: | wc -l
350
```

:joy_cat: 

But we are on it to fix some additional styling errors...

P.S.: What we really did was

```
$ git diff --files-only our_branch base | grep .rb | xargs rubocop
```

Commit:
Additional style fixes

myabc commented:
@NobodysNightmare very nice!

myabc commented:
but I think you mean:

```
git diff --files-only our_branch base | grep .rb | xargs rubocop -a
```

NobodysNightmare commented:
Nope, we went the extra mile to understand what we are doing ^^

ulferts commented on spec/lib/api/v3/priorities/priority_representer_spec.rb:
```diff
@@ -32,19 +32,43 @@
+      end
+      it 'should have an active flag' do
+        should have_json_path('isActive')
+      end
```
@Azure7111, @NobodysNightmare you could (and should) tests for the specific values. e.g. 

```
it 'should have an id' do
  is_expected.to be_json_eql(priority.id.to_json).at_path('id')
end
```

myabc replied:
(`is_expected.to` does the same thing as `should`, but probably reads slightly better)

ulferts commented on lib/api/v3/work_packages/work_package_representer.rb:
```diff
@@ -231,9 +239,16 @@ class WorkPackageRepresenter < ::API::Decorators::Single
                  },
                  setter: -> (value, *) { represented.description = value['raw'] },
                  render_nil: true
-        property :priority, getter: -> (*) { priority.try(:name) }, render_nil: true
```
@0xF013, @NobodysNightmare  this is going to destroy displaying the priority in the WP pane.

ulferts commented on lib/api/v3/utilities/path_helper.rb:
```diff
@@ -70,6 +70,10 @@ def self.priorities
           end
 
+          def self.priority(id)
+            "#{priorities}/#{id}"
```
This is untested code :scream:

-> spec/lib/api/v3/utilities/path_helper_spec.rb

NobodysNightmare replied:
not anymore ;)

ulferts commented on spec/requests/api/v3/priority_resource_spec.rb:
```diff
@@ -28,20 +28,22 @@
 
 require 'spec_helper'
 require 'rack/test'
+require 'pry'
```
@Azure7111, @NobodysNightmare 

No, it's not required.

ulferts commented on spec/requests/api/v3/priority_resource_spec.rb:
```diff
@@ -55,5 +57,61 @@
+        allow(User).to receive(:current).and_return current_user
+        member = FactoryGirl.build(:member, user: current_user, project: project)
+        member.role_ids = [role.id]
+        member.save!
```
@Azure7111 @NobodysNightmare 

Not an issue, just a pointer -> there is a more condensed way of writing this: https://github.com/opf/openproject/blob/dev/spec/models/query/results_spec.rb#L71

ulferts commented:
@Azure7111 @NobodysNightmare generally looking good.

Thanks for rubocopifying :balloon:

Please let me know who is going to address the details pane incompatibility issue and in which PR.

NobodysNightmare committed:
test for actual values

NobodysNightmare committed:
test new method in path helper

NobodysNightmare committed:
remove useless require

NobodysNightmare committed:
use more condensed way of creating users

- also remove requires that seem to have failed the status_resource_spec for me

NobodysNightmare committed:
return 403 for statuses when not authorized

this violated the spec and was found during refactoring...

ulferts commented:
One other thing I noticed when taking a look at the read part (a little late, I know) is that the priority is not embedded. That is inconsistent.

NobodysNightmare commented:
@ulferts @0xF013 I tried to look into the JS code in the hope to quickly identify where I needed to replace the access to `priority` with the access to `_links/priority/title`... however, I had no luck...

@0xF013 Would you mind fixing the WP pane? Do you need a work package for that? Would you do it in an extra PR?

0xF013 commented:
@NobodysNightmare I can do it. Should I hijack this branch?

NobodysNightmare commented:
You are welcome to do that :)

I will just add the missing embedding that @ulferts mentioned.

edit: I am just done with that. The branch is all yours now ;-)

NobodysNightmare committed:
also embed the priority

0xF013 committed:
change priority property source

0xF013 committed:
tests for priority change

ulferts commented:
Thanks everybody. Looks good.
